### PR TITLE
Library Version implementation.

### PR DESCRIPTION
**Pull request by vkryachko**
The change adds a mechanism for libraries, sdks to register their versions with the runtime.

### CODE REVIEW

File: firebase-common/src/main/java/com/google/firebase/platforminfo/PlatformInfoToken.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+// limitations under the License.
+
+package com.google.firebase.platforminfo;
+
+/** The class is not public to ensure other components cannot depend on it. */
+class PlatformInfoToken {
+  final String key;
+  final String value;
+
+  PlatformInfoToken(String key, String value) {
```
Comment (ashwinraghav) [inline firebase-common/src/main/java/com/google/firebase/platforminfo/PlatformInfoToken.java]: I feel like this needs some sort of namespacing per library for it to be PlatformInfo?
Is this intended to be just version or are we planning to start capturing other things in here?
```
PlatformInfoToken(String sdk, String key, String value) {
}
```

File: firebase-common/src/main/java/com/google/firebase/platforminfo/DefaultPlatformInfoSupplier.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+public class DefaultPlatformInfoSupplier implements PlatformInfoSupplier {
+  private final String tokens;
+
+  private DefaultPlatformInfoSupplier(Set<PlatformInfoToken> tokens) {
+    this.tokens = toUserAgent(tokens);
+  }
+
+  @Override
+  public String getUserAgent() {
+    if (PlatformInfoCompat.getGlobalInfo().isEmpty()) {
```
[the remark this answers is not in the document]
Comment (vkryachko) [reply firebase-common/src/main/java/com/google/firebase/platforminfo/DefaultPlatformInfoSupplier.java]: Yes

File: firebase-common/src/main/java/com/google/firebase/platforminfo/PlatformInfoToken.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.platforminfo;
+
+/** The class is not public to ensure other components cannot depend on it. */
+class PlatformInfoToken {
```
[the remark this answers is not in the document]
Comment (vkryachko) [reply firebase-common/src/main/java/com/google/firebase/platforminfo/PlatformInfoToken.java]: A few points
* I treat it as an implementation detail and so exposing it through components would mean teams can start depending on it.
* Depending on Set<PlatformInfoToken> is not very useful for anyone as it does not provide a full set(due to PlatformInfoCompat's existence), and so allowing SDKs to directly depend on it gives a false sense of completeness
* Additionally I don't want this mechanism to be (mis)used for inter-sdk communication

File: firebase-common/src/main/java/com/google/firebase/platforminfo/PlatformInfoToken.java (hunk 1 above)
Comment (vkryachko) [reply firebase-common/src/main/java/com/google/firebase/platforminfo/PlatformInfoToken.java]: There are no plans to capture anything other than sdk/library versions, so I feel like (key,value) pairs are sufficient to represent that?

File: firebase-common/src/main/java/com/google/firebase/platforminfo/SDKVersion.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+// limitations under the License.
+
+package com.google.firebase.platforminfo;
+
+import com.google.auto.value.AutoValue;
+import javax.annotation.Nonnull;
+
+/** The class is not public to ensure other components cannot depend on it. */
+@AutoValue
+abstract class SDKVersion {
```
Comment (vkryachko) [inline firebase-common/src/main/java/com/google/firebase/platforminfo/SDKVersion.java]: If we go with a version-related class name as opposed to PlatformInfo, then it might make sense to be consistent with the ios folks who call it `LibraryVersion`

File: firebase-common/src/main/java/com/google/firebase/platforminfo/SDKVersion.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  public static Builder builder() {
```
Comment (vkryachko) [inline firebase-common/src/main/java/com/google/firebase/platforminfo/SDKVersion.java]: I feel like a builder is overkill for a class that encodes 2 strings, can we have a plain `create(String, String)`?

File: firebase-common/src/main/java/com/google/firebase/FirebaseApp.java [hunk 6]
```diff
@@ -151,6 +153,8 @@
   @GuardedBy("LOCK")
   static final Map<String, FirebaseApp> INSTANCES = new ArrayMap<>();
 
+  public static final String FIREBASE_COMMON = "firebase-common";
```
Comment (vkryachko) [inline firebase-common/src/main/java/com/google/firebase/FirebaseApp.java]: s/public/private/ ?

File: firebase-common/src/main/java/com/google/firebase/FirebaseApp.java [hunk 7]
```diff
@@ -536,7 +540,9 @@ protected FirebaseApp(Context applicationContext, String name, FirebaseOptions o
             registrars,
             Component.of(applicationContext, Context.class),
             Component.of(this, FirebaseApp.class),
-            Component.of(options, FirebaseOptions.class));
+            Component.of(options, FirebaseOptions.class),
+            SDKVersionComponentFactory.createComponent(FIREBASE_COMMON, BuildConfig.VERSION_NAME),
```
Comment (vkryachko) [inline firebase-common/src/main/java/com/google/firebase/FirebaseApp.java]: optional nit: this seems a bit verbose and #enterprisy at the call site, would something like `SDKVersionComponent.create(String, String)` work?

File: firebase-common/src/main/java/com/google/firebase/platforminfo/GamesSDKVersionRegistrar.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+/**
+ * In order to allow the C++ and Unity SDKs to publish their versions without the use of the
+ * components framework, we have a mechanism where the versions can be wired as out of band as side
+ * effects. See {@link GamesSDKVersionRegistrar#registerVersion(String, String)}
+ *
+ * <p>Java libraries should use {@link SDKVersionComponentFactory#createComponent(String, String)}
+ * instead.
+ */
+public class GamesSDKVersionRegistrar {
```
Comment (vkryachko) [inline firebase-common/src/main/java/com/google/firebase/platforminfo/GamesSDKVersionRegistrar.java]: Might make sense to have a more generic name here and not encode the present use case in the name?

File: firebase-common/src/main/java/com/google/firebase/platforminfo/DefaultUserAgentPublisher.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    this.gamesSDKRegistrar = gamesSDKRegistrar;
+  }
+
+  /**
+   * Returns the user agent string that is computed as follows 1. For our JavaSDKs, the string is
+   * computed in advance since the components framework guarantees that we receive all published
+   * versions. 2. For our GamesSDKs, the strings are recomputed each time since the registration of
+   * the versions happens out of band and we take the optimistic approach of recomputing each time.
+   *
+   * @return
```
Comment (vkryachko) [inline firebase-common/src/main/java/com/google/firebase/platforminfo/DefaultUserAgentPublisher.java]: Can you remove `@return`s, here and throughout?

File: firebase-common/src/main/java/com/google/firebase/platforminfo/SDKVersionComponentFactory.java [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+/** Factory to create a component that publishes the version of an SDK */
+public class SDKVersionComponentFactory {
+  private SDKVersionComponentFactory() {}
+
+  /**
+   * Creates a component that publishes SDK versions
+   *
+   * @param sdkName
+   * @param version
+   * @return
```
Comment (vkryachko) [inline firebase-common/src/main/java/com/google/firebase/platforminfo/SDKVersionComponentFactory.java]: `@param`s don't seem to provide any value here as well

File: firebase-common/src/main/java/com/google/firebase/platforminfo/GamesSDKVersionRegistrar.java [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+  private final Set<SDKVersion> infos = new HashSet<>();
+  private static volatile GamesSDKVersionRegistrar instance;
+
+  private GamesSDKVersionRegistrar() {}
```
Comment (vkryachko) [inline firebase-common/src/main/java/com/google/firebase/platforminfo/GamesSDKVersionRegistrar.java]: Might make sense to open it up a bit(e.g. package-private) so that it can be used in tests below instead of having to use a mock there. wdyt?

File: firebase-common/src/test/java/com/google/firebase/platforminfo/DefaultUserAgentPublisherTest.java [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+import java.util.Set;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.annotation.Config;
+
+@RunWith(RobolectricTestRunner.class)
+@Config(manifest = "NoAutoDataCollectionAndroidManifest.xml")
+public class DefaultUserAgentPublisherTest {
```
Comment (vkryachko) [inline firebase-common/src/test/java/com/google/firebase/platforminfo/DefaultUserAgentPublisherTest.java]: Does this test have to be robolectric? can it be a plain unit test?

File: firebase-common/src/testUtil/com/google/firebase/components/VersionPublishingComponent.java [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.components;
+
+class VersionPublishingComponent {
```
Comment (vkryachko) [inline firebase-common/src/testUtil/com/google/firebase/components/VersionPublishingComponent.java]: Is this file needed? same for the one below (.swp)?

File: firebase-common/src/test/java/com/google/firebase/platforminfo/DefaultUserAgentPublisherTest.java (hunk 12 above)
Comment (ashwinraghav) [reply firebase-common/src/test/java/com/google/firebase/platforminfo/DefaultUserAgentPublisherTest.java]: Good point

File: firebase-common/src/main/java/com/google/firebase/platforminfo/GamesSDKVersionRegistrar.java (hunk 11 above)
Comment (ashwinraghav) [reply firebase-common/src/main/java/com/google/firebase/platforminfo/GamesSDKVersionRegistrar.java]: The mock was intended to be used only for the sake of isolation. Not much else. It did not hinder testability IMO

Comment (ashwinraghav) [reply firebase-common/src/main/java/com/google/firebase/platforminfo/GamesSDKVersionRegistrar.java]: Added tests too

File: firebase-common/src/main/java/com/google/firebase/platforminfo/GamesSDKVersionRegistrar.java [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+  /** Returns registered versions */
+  Set<LibraryVersion> getRegisteredVersions() {
+    synchronized (infos) {
+      return Collections.unmodifiableSet(infos);
+    }
+  }
+
+  /** Returns an instance of {@link GamesSDKVersionRegistrar} */
+  public static GamesSDKVersionRegistrar getInstance() {
+    if (instance == null) {
```
Comment (vkryachko) [inline firebase-common/src/main/java/com/google/firebase/platforminfo/GamesSDKVersionRegistrar.java]: You can make it even faster by assigning to a local variable first, see [1](search for localRef on the page)

[1] https://en.wikipedia.org/wiki/Double-checked_locking#Usage_in_Java

File: firebase-common/src/main/java/com/google/firebase/platforminfo/GamesSDKVersionRegistrar.java [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+/**
+ * In order to allow the C++ and Unity SDKs to publish their versions without the use of the
+ * components framework, we have a mechanism where the versions can be wired as out of band as side
+ * effects. See {@link GamesSDKVersionRegistrar#registerVersion(String, String)}
+ *
+ * <p>Java libraries should use {@link LibraryVersionComponent#create(String, String)} instead.
+ */
+public class GamesSDKVersionRegistrar {
+  private final Set<LibraryVersion> infos = new HashSet<>();
+  private static volatile GamesSDKVersionRegistrar instance;
```
Comment (vkryachko) [inline firebase-common/src/main/java/com/google/firebase/platforminfo/GamesSDKVersionRegistrar.java]: Please make variable name all caps `INSTANCE`.

File: firebase-common/src/main/java/com/google/firebase/platforminfo/GamesSDKVersionRegistrar.java [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+import java.util.Set;
+
+/**
+ * In order to allow the C++ and Unity SDKs to publish their versions without the use of the
+ * components framework, we have a mechanism where the versions can be wired as out of band as side
+ * effects. See {@link GamesSDKVersionRegistrar#registerVersion(String, String)}
+ *
+ * <p>Java libraries should use {@link LibraryVersionComponent#create(String, String)} instead.
+ */
+public class GamesSDKVersionRegistrar {
```
Comment (vkryachko) [inline firebase-common/src/main/java/com/google/firebase/platforminfo/GamesSDKVersionRegistrar.java]: Not a fan of this name, but I'll leave it up to you to decide. lgtm

Comment (ashwinraghav) [reply firebase-common/src/main/java/com/google/firebase/platforminfo/GamesSDKVersionRegistrar.java]: Renamed to OutOfBandVersionRegistrar

Comment (ashwinraghav) [reply firebase-common/src/main/java/com/google/firebase/platforminfo/GamesSDKVersionRegistrar.java]: Hope that seems reasonable
